In [1]:
import pandas as pd
import numpy as np
import os

# Importing relevant dataframe from last exercise.

In [2]:
path=r'C:\Users\Usha Bharati\Downloads\02-2024 Instacart Basket Analysis'

In [3]:
ords_prods_merge_derived=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge_derived.pkl'))

In [4]:
ords_prods_merge=pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge.pkl'))

## Q2 Aggregated mean of the “order_number” column grouped by “department_id” for entire dataframe

In [5]:
ords_prods_merge_derived.groupby('department_id')

In [6]:
ords_prods_merge_derived.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Q3 Since aggregate function is applied for entire dataframe, we see total of 21 departments and the average mean of department_id-21 is the highest with 22.9.

## Q4 Loyalty flag for existing customers

In [7]:
ords_prods_merge_derived['max_order'] = ords_prods_merge_derived.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\Usha Bharati\AppData\Local\Temp\ipykernel_64888\3298953043.py:1: FutureWarning: The provided callable <function amax at 0x0000019B6C4C2700> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge_derived['max_order'] = ords_prods_merge_derived.groupby(['user_id'])['order_number'].transform(np.max)


In [8]:
ords_prods_merge_derived.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid range product  Regularly busy  Regularly busy   
1  ...  Mid range product  Regularly busy  Regularly busy   
2  ...  Mid range product     Busiest day    Busiest days   
3  ...  Mid range product  Regularly busy    Slowest days   
4  ...  Mid range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag average_spending  \
0           Most Orders        32  Regular customer         6.935811   
1        Average Orders        32  Regular customer         6.935811   
2        Average Orders         5      New customer         7.930208   
3           Most Orders         3      New customer         4.972414   
4        Average Orders         3      New customer         4.972414   

   spending_flag median_days_since_prior_order     frequency_flag  
0    Low spender                           8.0  Frequent customer  
1    Low spender                           8.0  Frequent customer  
2    Low spender                           8.0  Frequent customer  
3    Low spender                           9.0  Frequent customer  
4    Low spender                           9.0  Frequent customer  

[5 rows x 23 columns]

In [9]:
ords_prods_merge_derived.loc[ords_prods_merge_derived['max_order']>40,'loyalty_flag']= 'Loyal customer'

In [10]:
ords_prods_merge_derived.loc[(ords_prods_merge_derived['max_order']<=40)&(ords_prods_merge_derived['max_order']>10),'loyalty_flag']= 'Regular customer'

In [11]:
ords_prods_merge_derived.loc[ords_prods_merge_derived['max_order']<=10,'loyalty_flag']= 'New customer'

In [12]:
ords_prods_merge_derived['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [13]:
ords_prods_merge_derived[['user_id', 'loyalty_flag', 'order_number','max_order']].head(60)

user_id      loyalty_flag  order_number  max_order
0       138  Regular customer            28         32
1       138  Regular customer            30         32
2       709      New customer             2          5
3       764      New customer             1          3
4       764      New customer             3          3
5       777  Regular customer            16         26
6       825      New customer             3          9
7       910  Regular customer            12         12
8      1052  Regular customer            10         20
9      1052  Regular customer            15         20
10     1379      New customer             1         10
11     1480    Loyal customer            36         43
12     1480    Loyal customer            38         43
13     1480    Loyal customer            39         43
14     1494  Regular customer             6         13
15     1494  Regular customer             7         13
16     1494  Regular customer             9         13
17     1540  Regular customer             4         28
18     1540  Regular customer             5         28
19     1540  Regular customer             6         28
20     1540  Regular customer             7         28
21     1540  Regular customer             8         28
22     1540  Regular customer             9         28
23     1540  Regular customer            14         28
24     1540  Regular customer            15         28
25     1540  Regular customer            18         28
26     1540  Regular customer            19         28
27     1540  Regular customer            20         28
28     1540  Regular customer            21         28
29     1540  Regular customer            22         28
30     1540  Regular customer            24         28
31     1540  Regular customer            25         28
32     1540  Regular customer            27         28
33     1540  Regular customer            28         28
34     1598      New customer             1          4
35     1598      New customer             2          4
36     1598      New customer             3          4
37     1647      New customer             1         10
38     1647      New customer             4         10
39     1647      New customer             6         10
40     1647      New customer             8         10
41     1647      New customer             9         10
42     2833      New customer             1         10
43     2850  Regular customer             7         23
44     2857  Regular customer            24         25
45     3010  Regular customer             5         13
46     3029      New customer             1          5
47     3395      New customer             4          6
48     3904      New customer             1          3
49     4006    Loyal customer             2         42
50     4006    Loyal customer             3         42
51     4006    Loyal customer             5         42
52     4006    Loyal customer             9         42
53     4006    Loyal customer            12         42
54     4122  Regular customer            11         40
55     4141      New customer             2          9
56     4343      New customer             1          4
57     4553      New customer             3          4
58     4845  Regular customer             6         14
59     6866      New customer             1          6

# Q5

In [14]:
ords_prods_merge_derived.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## Loyal customers spend less compared to Regular and New customers. New customers spend the most, then regular customer and then loyal customers.

## Q6  Spending flag for each user based on the average price

In [15]:
ords_prods_merge_derived['average_spending'] = ords_prods_merge_derived.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Usha Bharati\AppData\Local\Temp\ipykernel_64888\3530399239.py:1: FutureWarning: The provided callable <function mean at 0x0000019B6C4C3420> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge_derived['average_spending'] = ords_prods_merge_derived.groupby(['user_id'])['prices'].transform(np.mean)


In [16]:
ords_prods_merge_derived.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid range product  Regularly busy  Regularly busy   
1  ...  Mid range product  Regularly busy  Regularly busy   
2  ...  Mid range product     Busiest day    Busiest days   
3  ...  Mid range product  Regularly busy    Slowest days   
4  ...  Mid range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag average_spending  \
0           Most Orders        32  Regular customer         6.935811   
1        Average Orders        32  Regular customer         6.935811   
2        Average Orders         5      New customer         7.930208   
3           Most Orders         3      New customer         4.972414   
4        Average Orders         3      New customer         4.972414   

   spending_flag median_days_since_prior_order     frequency_flag  
0    Low spender                           8.0  Frequent customer  
1    Low spender                           8.0  Frequent customer  
2    Low spender                           8.0  Frequent customer  
3    Low spender                           9.0  Frequent customer  
4    Low spender                           9.0  Frequent customer  

[5 rows x 23 columns]

In [17]:
ords_prods_merge_derived.loc[ords_prods_merge_derived['average_spending']<10,'spending_flag']= 'Low spender'

In [18]:
ords_prods_merge_derived.loc[ords_prods_merge_derived['average_spending']>=10,'spending_flag']= 'High spender'

In [19]:
ords_prods_merge_derived['spending_flag'].value_counts(dropna=False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [20]:
pd.options.display.max_rows=None

In [21]:
ords_prods_merge_derived[['user_id','prices','average_spending','spending_flag']].head(100)

user_id  prices  average_spending spending_flag
0       138     5.8          6.935811   Low spender
1       138     5.8          6.935811   Low spender
2       709     5.8          7.930208   Low spender
3       764     5.8          4.972414   Low spender
4       764     5.8          4.972414   Low spender
5       777     5.8          6.935398   Low spender
6       825     5.8          5.957576   Low spender
7       910     5.8          6.680000   Low spender
8      1052     5.8          7.162500   Low spender
9      1052     5.8          7.162500   Low spender
10     1379     5.8          5.813636   Low spender
11     1480     5.8          5.081633   Low spender
12     1480     5.8          5.081633   Low spender
13     1480     5.8          5.081633   Low spender
14     1494     5.8          6.803797   Low spender
15     1494     5.8          6.803797   Low spender
16     1494     5.8          6.803797   Low spender
17     1540     5.8          6.607122   Low spender
18     1540     5.8          6.607122   Low spender
19     1540     5.8          6.607122   Low spender
20     1540     5.8          6.607122   Low spender
21     1540     5.8          6.607122   Low spender
22     1540     5.8          6.607122   Low spender
23     1540     5.8          6.607122   Low spender
24     1540     5.8          6.607122   Low spender
25     1540     5.8          6.607122   Low spender
26     1540     5.8          6.607122   Low spender
27     1540     5.8          6.607122   Low spender
28     1540     5.8          6.607122   Low spender
29     1540     5.8          6.607122   Low spender
30     1540     5.8          6.607122   Low spender
31     1540     5.8          6.607122   Low spender
32     1540     5.8          6.607122   Low spender
33     1540     5.8          6.607122   Low spender
34     1598     5.8          7.123529   Low spender
35     1598     5.8          7.123529   Low spender
36     1598     5.8          7.123529   Low spender
37     1647     5.8          4.853521   Low spender
38     1647     5.8          4.853521   Low spender
39     1647     5.8          4.853521   Low spender
40     1647     5.8          4.853521   Low spender
41     1647     5.8          4.853521   Low spender
42     2833     5.8          7.377143   Low spender
43     2850     5.8          8.054153   Low spender
44     2857     5.8          8.353793   Low spender
45     3010     5.8          6.452800   Low spender
46     3029     5.8          7.960606   Low spender
47     3395     5.8          6.322222   Low spender
48     3904     5.8          7.511111   Low spender
49     4006     5.8          7.441100   Low spender
50     4006     5.8          7.441100   Low spender
51     4006     5.8          7.441100   Low spender
52     4006     5.8          7.441100   Low spender
53     4006     5.8          7.441100   Low spender
54     4122     5.8          7.598765   Low spender
55     4141     5.8          6.033333   Low spender
56     4343     5.8          7.393939   Low spender
57     4553     5.8          5.715789   Low spender
58     4845     5.8          7.229630   Low spender
59     6866     5.8          4.807143   Low spender
60     7275     5.8          4.620833   Low spender
61     7669     5.8          9.106757   Low spender
62     7669     5.8          9.106757   Low spender
63     7669     5.8          9.106757   Low spender
64     7981     5.8          7.115714   Low spender
65     7981     5.8          7.115714   Low spender
66     7981     5.8          7.115714   Low spender
67     7981     5.8          7.115714   Low spender
68     7981     5.8          7.115714   Low spender
69     8703     5.8          6.593459   Low spender
70     8723     5.8          5.677397   Low spender
71     8735     5.8          5.850000   Low spender
72     9293     5.8         11.857143  High spender
73     9293     5.8         11.857143  High spender
74     9467     5.8          7.709091   Low spender
75     9694     5.8          7.075354   Low spender


## Q7 Frequency flag acc to median in “days_since_prior_order” column

In [22]:
ords_prods_merge_derived['median_days_since_prior_order'] = ords_prods_merge_derived.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Usha Bharati\AppData\Local\Temp\ipykernel_64888\4107362169.py:1: FutureWarning: The provided callable <function median at 0x0000019B6C5FE020> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge_derived['median_days_since_prior_order'] = ords_prods_merge_derived.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [23]:
ords_prods_merge_derived.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid range product  Regularly busy  Regularly busy   
1  ...  Mid range product  Regularly busy  Regularly busy   
2  ...  Mid range product     Busiest day    Busiest days   
3  ...  Mid range product  Regularly busy    Slowest days   
4  ...  Mid range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag average_spending  \
0           Most Orders        32  Regular customer         6.935811   
1        Average Orders        32  Regular customer         6.935811   
2        Average Orders         5      New customer         7.930208   
3           Most Orders         3      New customer         4.972414   
4        Average Orders         3      New customer         4.972414   

   spending_flag median_days_since_prior_order     frequency_flag  
0    Low spender                           8.0  Frequent customer  
1    Low spender                           8.0  Frequent customer  
2    Low spender                           8.0  Frequent customer  
3    Low spender                           9.0  Frequent customer  
4    Low spender                           9.0  Frequent customer  

[5 rows x 23 columns]

In [24]:
ords_prods_merge_derived.loc[ords_prods_merge_derived['median_days_since_prior_order']<=10,'frequency_flag']= 'Frequent customer'

In [25]:
ords_prods_merge_derived.loc[(ords_prods_merge_derived['median_days_since_prior_order']>10) & (ords_prods_merge_derived['median_days_since_prior_order']<=20),'frequency_flag']= 'Regular customer'

In [26]:
ords_prods_merge_derived.loc[ords_prods_merge_derived['median_days_since_prior_order']>20,'frequency_flag']= 'Non-frequent customer'

In [27]:
ords_prods_merge_derived['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

In [28]:
ords_prods_merge_derived.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid range product  Regularly busy  Regularly busy   
1  ...  Mid range product  Regularly busy  Regularly busy   
2  ...  Mid range product     Busiest day    Busiest days   
3  ...  Mid range product  Regularly busy    Slowest days   
4  ...  Mid range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag average_spending  \
0           Most Orders        32  Regular customer         6.935811   
1        Average Orders        32  Regular customer         6.935811   
2        Average Orders         5      New customer         7.930208   
3           Most Orders         3      New customer         4.972414   
4        Average Orders         3      New customer         4.972414   

   spending_flag median_days_since_prior_order     frequency_flag  
0    Low spender                           8.0  Frequent customer  
1    Low spender                           8.0  Frequent customer  
2    Low spender                           8.0  Frequent customer  
3    Low spender                           9.0  Frequent customer  
4    Low spender                           9.0  Frequent customer  

[5 rows x 23 columns]

## Exporting dataframe as pickle file

In [29]:
ords_prods_merge_derived.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_derived.pkl'))

In [30]:
ords_prods_merge_derived.dtypes

product_id                         int64
product_name                      object
aisle_id                           int64
department_id                      int64
prices                           float64
order_id                           int64
user_id                            int64
order_number                       int64
orders_day_of_week                 int64
order_hour_of_day                  int64
days_since_prior_order           float64
add_to_cart_order                  int64
reordered                          int64
price_range_loc                   object
busiest_day                       object
busiest_days                      object
busiest_period_of_day             object
max_order                          int64
loyalty_flag                      object
average_spending                 float64
spending_flag                     object
median_days_since_prior_order    float64
frequency_flag                    object
dtype: object